src/server.c

```diff
@@ -5938,6 +5945,58 @@ void linuxMemoryWarnings(void) {
     }
 }
 
+void linuxTimeWarnings(void) {
+#ifdef HAVE_PROC_STAT
+#define PROC_STAT_STIME_IDX 15
+    long long sys_time_ticks, sys_time_ticks_start;
+    unsigned long systime_us, system_hz, test_time_us;
+    struct timespec ts;
+    unsigned long long start_us;
+    system_hz = sysconf(_SC_CLK_TCK);
+
+    if (!get_proc_stat_ll(PROC_STAT_STIME_IDX, &sys_time_ticks_start))
+        return;
+    if (clock_gettime(CLOCK_MONOTONIC, &ts) < 0) {
+        return;
+    }
+    start_us = (ts.tv_sec * 1000000 + ts.tv_nsec / 1000);
+
+    /* clock_gettime() busy loop of 5 times system hz (for a system_hz of 100 this is 50ms).
+     * If our clocksource is configured correctly (vdso) this will result in no system calls.
+     * If our clocksource is inefficient it'll waste most of the busy loop in the kernel. */
```

Comment:
@yossigo are you ok with that?
50ms busy loop by default on any redis startup on linux?
there would be a config to supress the warning, which will also suppress the test, but still, i suppose that someone starting up many instance of redis by some script will be hit by that.

Reply:
I have to set I don't like that - not the busy loop and not the heuristics we have to apply that may also lead to false positives.

Reply:
I wouldn't define that as heuristics.. That's the real right way to test.. I.e. We don't care how it's configured. But rather the fact it uses "system" time. 
Note that the threshold for telling them apart is very clear.. I just wish there was a way to do that with just one call and not a loop

Reply:
@oranagra I think it is a heuristic, because we're comparing wall clock time with CPU stime.

Reply:
The check is basically built out of 3 steps:
1. Read stime from proc fs.
2. Busy loop longer than a system tick calling `clock_gettime()`.
3. Read stime from proc again.
Steps 1 and 3 are by definition a system call. Step 2 is either a system call on an un-optimized system or a a vdso call or an optimized system.
The _theoretical_ logic of the test is simple and not heuristic: if this is an un-optimized system, the stime diff will be larger than 0. In _practice_ since steps 1 and 3 perform system calls to read from proc fs then there's the slight chance that the stime diff will be affected by the procfs reads. This is what's heuristic about the test, but there's a very safe margin here.

On the other hand it surprised my that the un-optimized xen implementation spent only ~30% of the time in the system call. So I'm not totally sure what's happening there. But I'm sure it's due to, at least partially, not going through vdso on all calls to `clock_gettime()` which is enough to degrade performance.

Reply:
Decided to put this test under the `--check-system` mode instead of on startup.